c++/src/kj/async-io.c++

```diff
@@ -185,6 +186,922 @@ Maybe<Promise<uint64_t>> AsyncOutputStream::tryPumpFrom(
+    }
+
+    Promise<size_t> tryRead(void* readBuffer, size_t minBytes, size_t maxBytes) override {
+      KJ_FAIL_REQUIRE("can't read() again until previous pumpTo() completes");
+    }
+    Promise<uint64_t> pumpTo(AsyncOutputStream& output, uint64_t amount) override {
+      KJ_FAIL_REQUIRE("can't read() again until previous pumpTo() completes");
+    }
+
+    void abortRead() override {
+      canceler.cancel("abortRead() was called");
+      fulfiller.reject(KJ_EXCEPTION(FAILED, "read end of pipe was aborted"));
+      pipe.endState(*this);
+      pipe.abortRead();
+    }
+
+    Promise<void> write(const void* writeBuffer, size_t size) override {
+      KJ_REQUIRE(canceler.isEmpty(), "already pumping");
+
+      auto actual = kj::min(amount - pumpedSoFar, size);
+      return canceler.wrap(output.write(writeBuffer, actual)
+          .then([this,size,actual,writeBuffer]() -> kj::Promise<void> {
+        canceler.release();
+        pumpedSoFar += actual;
+
+        KJ_ASSERT(pumpedSoFar <= amount);
+        KJ_ASSERT(actual <= size);
+
+        if (pumpedSoFar == amount) {
+          // Done with pump.
+          fulfiller.fulfill(kj::cp(pumpedSoFar));
+          pipe.endState(*this);
+        }
+
+        if (actual == size) {
+          return kj::READY_NOW;
+        } else {
+          KJ_ASSERT(pumpedSoFar == amount);
+          return pipe.write(reinterpret_cast<const byte*>(writeBuffer) + actual, size - actual);
+        }
+      }));
+    }
+
+    Promise<void> write(ArrayPtr<const ArrayPtr<const byte>> pieces) override {
+      KJ_REQUIRE(canceler.isEmpty(), "already pumping");
+
+      size_t size = 0;
+      size_t needed = amount - pumpedSoFar;
+      for (auto i: kj::indices(pieces)) {
+        if (pieces[i].size() > needed) {
+          // The pump ends in the middle of this write.
+
+          auto promise = output.write(pieces.slice(0, i));
+
+          if (needed > 0) {
+            // The pump includes part of this piece, but not all. Unfortunately we need to split
+            // writes.
+            auto partial = pieces[i].slice(0, needed);
+            promise = promise.then([this,partial]() {
+              return output.write(partial.begin(), partial.size());
+            });
+            auto partial2 = pieces[i].slice(needed, pieces[i].size());
+            promise = canceler.wrap(promise.then([this,partial2]() {
+              canceler.release();
+              fulfiller.fulfill(kj::cp(amount));
+              pipe.endState(*this);
+              return pipe.write(partial2.begin(), partial2.size());
+            }));
+            ++i;
+          } else {
+            // The pump ends exactly at the end of a piece, how nice.
+            promise = canceler.wrap(promise.then([this]() {
+              canceler.release();
+              fulfiller.fulfill(kj::cp(amount));
+              pipe.endState(*this);
+            }));
+          }
+
+          auto remainder = pieces.slice(i + 1, pieces.size());
```

Comment:
Is this `i + 1` already accounted for with the `++i` above?

If we enter this function for the first time with `amount == 1` and `pieces` being a 2-element array of one-byte arrays, so:

```
pieces.size() == 2
pieces[i].size() == 1 for all i
amount == 1
pumpedSoFar == 0
```

Then the for loop will decrement `needed` to zero when `i == 0`, then when `i == 1`:

1. Write `pieces.slice(0, 1)`
2. Skip the split write, because the pump ends exactly at the end of a piece.
3. `remainder = pieces.slice(2, 2)` (an empty array)
4. Return without having written `piece[1]`.

Here's a test that I think expresses it:

```c++
KJ_TEST("Userland pipe pumpTo less than write amount") {
  kj::EventLoop loop;
  WaitScope ws(loop);

  auto pipe = newOneWayPipe();
  auto pipe2 = newOneWayPipe();
  auto pumpPromise = pipe.in->pumpTo(*pipe2.out, 1);

  auto pieces = kj::heapArray<ArrayPtr<const byte>>(2);
  byte a[1] = { 'a' };
  byte b[1] = { 'b' };
  pieces[0] = arrayPtr(a, 1);
  pieces[1] = arrayPtr(b, 1);

  auto writePromise = pipe.out->write(pieces);
  KJ_EXPECT(!writePromise.poll(ws));

  expectRead(*pipe2.in, "a").wait(ws);
  KJ_EXPECT(pumpPromise.wait(ws) == 1);
  KJ_EXPECT(!writePromise.poll(ws));

  pumpPromise = pipe.in->pumpTo(*pipe2.out, 1);

  expectRead(*pipe2.in, "b").wait(ws);
  KJ_EXPECT(pumpPromise.wait(ws) == 1);
  writePromise.wait(ws);
}
```

Assuming I'm not totally confused, a similar bug should exist with the split-write case.